dbms/src/Interpreters/ExpressionActions.cpp

```diff
@@ -562,6 +563,74 @@ std::string ExpressionActions::getSmallestColumn(const NamesAndTypesList & colum
     return res;
 }
 
+void ExpressionActions::pruneInputColumns(const Names & output_columns)
```

Comment:
What is the difference from `finalize`?

Reply:
It does not remove column, if a column is removed, it will cause error for nullaware join since it intends to reuse the columns: https://github.com/pingcap/tiflash/blob/adf85696270f864c0317526919b37fbdfe5a6417/dbms/src/Interpreters/NullAwareSemiJoinHelper.cpp#L322-L327

Reply:
If a column is required, why not set it to required_column?

Reply:
Because at the time of calling `pruneInputColumns`, we don't know which columns need to be kept.
For example, the original output columns are
`a, b, c, func(a), func(b), func(a) + func(b)`
And the output columns is `func(a) + func(b)`
If use `finalize`, then `finalize` will insert `REMOVE_COLUMN` to remove `func(a), func(b), a, b, c` to make sure the output column is `func(a) + func(b)`
While for nullaware join, it actually require the output block keeps `a, b` as the first/second column. So not easy to reuse `finalize` function.